### PR TITLE
oop-static_modifier-hard-q6

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
Please see my feedback on other questions in this category and apply it to this PR.

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+- Question Type: Java question
+
+## Farm Class
+
+Your `Farm` class must have the following requirements:
+
+- Attributes:
+    - A _static_ `double` called **foodDonation**. This represents the total number of food available for
+      donation to `Farm` objects, and should be set initially to 50 without having to call the constructor. Note that
+      it should not be visible from other classes.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:12]:
```suggestion
    - A _static_ `double` called **foodInSilo**. This represents the total number of food available in the silo for 
    `Farm` objects, and should be set initially to 50 without having to call the constructor. Note that
      it should not be visible from other classes.
```

Comment (SamamthaH) [reply src/oop/static_modifier/hard/q6/QuestionText.md:12]: change attribute name in the rest accordingly

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    - A constructor that is accessible from other methods, and accepts a `double` and an `int` as parameters to set the
+      **amountPerAnimal** and **numAnimals** attribute of the object.
+
+- Methods:
+    - Note that all methods in this class should be visible from another class
+    - A `foodNeeded` method which returns a `double` equal to the value of **amountPerAnimal** times **numAnimals**
+      which represents the total amount of food needed by the `Farm` object.
+    - A `receiveDonation` method which prints "Thank you for your donation." if the total food needed is less than the
+      amount available for donation, **foodDonation** value and subtract it from available food, and prints "Thank you,
+      but there is not enough for everyone here." otherwise.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:28]:
```suggestion
    - A `feedAnimals` method which prints "Animals fed and happy." if the total food needed is less than the
      amount available in the **foodInSilo** variable, and subtract it from available food. If there is not enough
      food it should print "We need to buy more food." and subtract nothing.
```

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## TestDonation Class
+
+Your `TestDontion` class must have the following items:
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:32]:
```suggestion
## TestFeeding Class

Your `TestFeeding` class must have the following items:
```

File: src/oop/static_modifier/hard/q6/QuestionText.md (hunk 2 above)
Comment (SamamthaH) [reply src/oop/static_modifier/hard/q6/QuestionText.md:28]: update other areas a relevant

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+- Main method:
+    - A `main` method which initializes 3 `Farm` objects with the following **amountPerAnimal** amounts in order:
+      0.5, 0.2, 1 and following **numAnimals** numbers in order: 10, 20, 50. Then in a loop from first to third farm
+      it should try to donate food by printing "Try to donate here... " and using **recieveDonation** method until there
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:37]:
```suggestion
      it should try to donate food by printing "Try to feed animals here... " and using **recieveDonation** method until there
```

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    public void farmHasRequiredConstructor() {
+        String missingConstructorMessage = "The %s class is missing a required constructor. Make sure that it is named correctly and has the correct parameters.";
+        String wrongAccessModifier = "The %s class constructor has the wrong visibility modifier. Make sure that it is visible from all other classes.";
+        Class<?>[] constructorArgs = new Class[]{double.class, int.class};
+        assertTrue(testClass.hasConstructor(constructorArgs), String.format(missingConstructorMessage, className));
+        assertTrue(testClass.hasModifier(constructorArgs, "public"), String.format(wrongAccessModifier, className));
+    }
+
+    @Test
+    @Order(3)
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:73]: make this the first test

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                {value2, int.class}
+        };
+        Object checkupInstance = testClass.createInstance(constructorArgs);
+        _assertEquals(value1, testClass.getFieldValue(checkupInstance, attributeName1), String.format(wrongValueMessage, className, attributeName1));
+        _assertEquals(value2, testClass.getFieldValue(checkupInstance, attributeName2), String.format(wrongValueMessage, className, attributeName2));
+    }
+
+    @Test
+    @Order(5)
+    public void receiveDonationIsDefinedCorrectly() {
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:112]: this test is no longer needed

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        Object[][] constructorArgs = {
+                {value1, double.class},
+                {value2, int.class}
+        };
+        Object classInstance = testClass.createInstance(constructorArgs);
+        testClass.setFieldValue(null, value3, staticAttributeName);
+        Clause[] testSentence;
+        if (value1 * value2 <= value3) {
+            testSentence = new Clause[]{
+                    new StringLiteral("Thank you for your donation."),
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:141]: make these strings variables at the top of the tests

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        } else {
+            _assertEquals(value3, testClass.getFieldValue(null, staticAttributeName),
+                    "Your " + methodName2 + " method changes the value of " + staticAttributeName + " attribute when it should not.");
+        }
+        testClass.setFieldValue(null, initialBeams, staticAttributeName);
+    }
+
+    @Test
+    @Order(7)
+    public void foodNeededIsDefinedCorrectly() {
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:163]: move the contents of this test into  farmClassHasCorrectFoodNeededMethod

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+                    new StringLiteral("Thank you for your donation."),
+                    new Optional(new StringLiteral(" "))
+            };
+        } else {
+            testSentence = new Clause[]{
+                    new StringLiteral("Thank you, but there is not enough for everyone here."),
+                    new Optional(new StringLiteral(" "))
+            };
+        }
+        testClass.callMethod(methodName2, classInstance, testSentence);
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:150]: add an error message

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+                "Your " + methodName1 + " method does not return the correct value.");
+    }
+
+    @Test
+    @Tag("dependent1")
+    @Order(9)
+    public void correctMainMethod() throws Throwable {
+        double temp = initialBeams;
+        Clause[] clauses = new Clause[8];
+        if (temp >= 0.5 * 10) {
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:193]: store the numeric values as variables at the top of the tests

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+# oop-static_modifier-hard-q6
+
+- Question Type: Java question
+
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:4]: Animals are an important part of farm life and require a lot of food. Write a Java class called `Farm` to keep track of the animals and available food using the requirements below. Then write a class called `TestFeeding` to test the `Farm` using the next set of requirements below.

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+## TestDonation Class
+
+Your `TestDontion` class must have the following items:
+
+- Main method:
+    - A `main` method which initializes 3 `Farm` objects with the following **amountPerAnimal** amounts in order:
+      0.5, 0.2, 1 and following **numAnimals** numbers in order: 10, 20, 50. Then in a loop from first to third farm
+      it should try to donate food by printing "Try to donate here... " and using **recieveDonation** method until there
+      can not be any food donated. Then it should print "There is no more food left today.". All prints are in separate lines.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:38]: It does not currently print "There is no more food left today."

Comment (SamamthaH) [reply src/oop/static_modifier/hard/q6/QuestionText.md:38]:
```suggestion
      can not be any food donated. Then it should print "There is no more food left today." All prints are in separate lines.
```

File: src/oop/static_modifier/hard/q6/MainTest.java (hunk 7 above)
Comment (SamamthaH) [reply src/oop/static_modifier/hard/q6/MainTest.java:141]: The period in it should also be escaped since periods have a special meaning in regex

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See Sam's review

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+## TestFeeding Class
+
+Your `TestFeeding` class must have the following items:
+
+- Main method:
+    - A `main` method which initializes 3 `Farm` objects with the following **amountPerAnimal** amounts in order:
+      0.5, 0.2, 1 and following **numAnimals** numbers in order: 10, 20, 50. Then in a loop from first to third farm
+      it should try to donate food by printing "Try to feed animals here... " and using **feedAnimals** method until there
+      can not be any food donated. Then it should print "There is no more food left today." All prints are in separate lines.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:38]: change TestFeeding to match this description. You are still not printing "There is no more food left today." or "Try to feed animals here..."

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+        Object checkupInstance = testClass.createInstance(constructorArgs);
+        _assertEquals(value1, testClass.getFieldValue(checkupInstance, attributeName1), String.format(wrongValueMessage, className, attributeName1));
+        _assertEquals(value2, testClass.getFieldValue(checkupInstance, attributeName2), String.format(wrongValueMessage, className, attributeName2));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependency1")
+    @Order(6)
+    public void farmClassHasCorrectReceiveDonationMethod(double value1, int value2, double value3) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:123]: fix this name

File: src/oop/static_modifier/hard/q6/QuestionText.md (hunk 11 above)
Comment (SamamthaH) [reply src/oop/static_modifier/hard/q6/QuestionText.md:4]: Add this as question text

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 15]
```diff
@@ -0,0 +1,38 @@
+# oop-static_modifier-hard-q6
+
+- Question Type: Java question
+
+## Farm Class
+
+Your `Farm` class must have the following requirements:
+
+- Attributes:
+    - A _static_ `double` called **foodInSilo**. This represents the total number of food available in the silo for
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:10]:
```suggestion
    - A _static_ `double` called **foodInSilo**. This represents the total amount of food available in the silo for
```

File: src/oop/static_modifier/hard/q6/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+    public void farmConstructorInitializesValuesCorrectly(double value1, int value2) throws Throwable {
+        String wrongValueMessage = "The %s constructor did not initialize the %s attribute to the correct value based on the parameters passed to the constructor.";
+        Object[][] constructorArgs = {
+                {value1, double.class},
+                {value2, int.class}
+        };
+        Object checkupInstance = testClass.createInstance(constructorArgs);
+        _assertEquals(value1, testClass.getFieldValue(checkupInstance, attributeName1), String.format(wrongValueMessage, className, attributeName1));
+        _assertEquals(value2, testClass.getFieldValue(checkupInstance, attributeName2), String.format(wrongValueMessage, className, attributeName2));
+    }
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/MainTest.java:117]: add a test that checks they don't alter the static value on calling the constructor

File: src/oop/static_modifier/hard/q6/QuestionText.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+- Constructors:
+    - A constructor that is accessible from other methods, and accepts a `double` and an `int` as parameters to set the
+      **amountPerAnimal** and **numAnimals** attribute of the object.
+
+- Methods:
+    - Note that all methods in this class should be visible from another class
+    - A `foodNeeded` method which returns a `double` equal to the value of **amountPerAnimal** times **numAnimals**
+      which represents the total amount of food needed by the `Farm` object.
+    - A `feedAnimals` method which prints "Animals fed and happy" if the total food needed is less than the
+      amount available in the **foodInSilo** variable, and subtract it from available food. If there is not enough
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q6/QuestionText.md:27]:
```suggestion
      amount available in the **foodInSilo** variable, and subtracts the value from the available food. If there is not enough
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Pull request merged**